File: leetcode-04/solutions/python3/0962.py
class Solution:
    def maxWidthRamp(self, A: List[int]) -> int:
        ans = 0
        stack = []

        for i, a in enumerate(A):
            if not stack or a < A[stack[-1]]:
                stack.append(i)

        for j in range(len(A))[::-1]:
            while stack and A[stack[-1]] <= A[j]:
                ans = max(ans, j - stack.pop())

        return ans

File: leetcode-04/solutions/cpp/0962.cpp
class Solution {
 public:
  int maxWidthRamp(vector<int>& A) {
    const int n = A.size();

    int ans = 0;
    stack<int> stack;

    for (int i = 0; i < n; i++)
      if (stack.empty() || A[i] < A[stack.top()]) stack.push(i);

    for (int j = n - 1; j > ans; j--)
      while (!stack.empty() && A[stack.top()] <= A[j]) {
        ans = max(ans, j - stack.top());
        stack.pop();
      }

    return ans;
  }
};